mongo/collection.go

```diff
@@ -400,27 +400,19 @@ func (coll *Collection) insert(ctx context.Context, documents []interface{},
 //
 // For more information about the command, see https://www.mongodb.com/docs/manual/reference/command/insert/.
 func (coll *Collection) InsertOne(ctx context.Context, document interface{},
-	opts ...*options.InsertOneOptions) (*InsertOneResult, error) {
+	opts ...Options[options.InsertOneArgs]) (*InsertOneResult, error) {
```

Comment:
@matthewdale @qingyang-hu Are we okay with the fact that this pattern is not-breaking for one instance of opts, but is a breaking change for a slice of opts?

```go
coll.InsertOne(ctx, bson.D{{"x", 1}}, options.InsertOne())
```

```go
// This is a breaking change
iopts := []*options.InsertOneOptions{
	options.InsertOne(),
	options.InsertOne(),
}

coll.InsertOne(ctx, bson.D{{"x", 1}}, iopts...)
```

To resolve the breaking change, you have to update the slice type to use mongo.Options:
```go
// This is a breaking change
iopts := []*mongo.Options[options.InsertOneArgs]{
	options.InsertOne(),
	options.InsertOne(),
}

coll.InsertOne(ctx, bson.D{{"x", 1}}, iopts...)
```

Reply:
Here are the use cases I know of where that will be a breaking change:
* Any package that wraps the `Collection` APIs and passes through options.
  * ODMs, like [mgm](https://github.com/Kamva/mgm). E.g. see [here](https://github.com/Kamva/mgm/blob/aff6e8d7b45bd3852c2b2e9fc1dfd87cd464f835/collection.go#L49).
  * Some projects have internal APIs that wrap the `Collection` API.
* Interfaces that describe the `Collection` API.

See a Sourcegraph search for all known public code that would be impacted [here](https://sourcegraph.com/search?q=context:global+lang:Go++-repo:%5Egithub%5C.com/mongodb/mongo-go-driver%24+repo:has.content%28go.mongodb.org/mongo-driver%29+-file:%28%5E%7C/%29vendor/+%28%5C%5B%5C%5D%7C%5C.%5C.%5C.%29%5C*options%5C.%5Ba-zA-Z%5D%2BOptions&patternType=regexp&case=yes&sm=0&groupBy=repo). There are matches in about 140 repos.

The change is straightforward and seems like it can mostly be accomplished with a find & replace, so it doesn't seem like a major breaking change.